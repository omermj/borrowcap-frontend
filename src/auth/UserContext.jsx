import React from "react";

/** Context to provide currentUser thoughout the app */

const UserContext = React.createContext();

export default UserContext;
