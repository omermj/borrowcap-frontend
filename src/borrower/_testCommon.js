import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = () => {
  // Mock api call with active requests
  BorrowcapApi.getActiveRequestsByBorrowerId.mockResolvedValue([
    {
      id: 4,
      amtRequested: 10000,
      purpose: "Business",
      appOpenDate: "2021/01/01",
      interestRate: 0.05,
      term: 12,
      installmentAmt: 1000,
    },
    {
      id: 5,
      amtRequested: 20000,
      purpose: "Home",
      appOpenDate: "2021/02/01",
      interestRate: 0.07,
      term: 24,
      installmentAmt: 2000,
    },
  ]);

  BorrowcapApi.getApprovedRequestsByUserId.mockResolvedValue({
    borrower: [
      {
        id: 2,
        amtRequested: "13000",
        amtApproved: "12000",
        amtFunded: "6000",
        purpose: "Education",
        appOpenDate: "2023-01-01",
        appApprovedDate: "2023/01/05",
        fundingDeadline: "2023/02/05",
        interestRate: "0.07",
        term: 24,
        installmentAmt: "537.27",
        availableForFunding: false,
      },
    ],
  });

  BorrowcapApi.getApprovedRequest.mockResolvedValue({
    id: 8,
    borrowerId: 2,
    amtRequested: "13000",
    amtApproved: "12000",
    amtFunded: "6000",
    purpose: "Education",
    purposeId: 3,
    appOpenDate: "2023-01-01",
    appApprovedDate: "2023/01/05",
    fundingDeadline: "2023/02/05",
    interestRate: "0.07",
    term: 24,
    installmentAmt: "537.27",
    availableForFunding: true,
    isFunded: false,
    annualIncome: "100000",
    otherMonthlyDebt: "1200",
  });

  BorrowcapApi.enableFundingForApprovedRequest.mockResolvedValue({
    message: "Funding is enabled",
  });

  BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    borrower: [
      {
        id: 6,
        amtFunded: "25000",
        fundedDate: "2023/01/05",
        interestRate: "0.07",
        term: 36,
        installmentAmt: "771.93",
        remainingBalance: "25000",
      },
    ],
  });
};

module.exports = { testUser, commonBeforeEach };
