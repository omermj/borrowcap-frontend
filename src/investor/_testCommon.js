import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getFundedLoansByUserId.mockResolvedValue({
    investor: [
      {
        id: 1,
        investorId: 3,
        amtInvested: "5000",
        amtFunded: "5000",
        fundedDate: "2023/01/01",
        interestRate: "0.05",
        term: 12,
        installmentAmt: "428.04",
        remainingBalance: "5000",
      },
    ],
  });
  await BorrowcapApi.getAvailableInvestments.mockResolvedValue([
    {
      id: 7,
      amtApproved: "8000",
      amtFunded: "0",
      purpose: "Medical",
      approvedDate: "2023/01/01",
      fundingDeadline: "2023/02/1",
      interestRate: "0.07",
      term: 24,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
