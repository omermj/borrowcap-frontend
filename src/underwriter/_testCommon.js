import BorrowcapApi from "../api/api";

// Mock user
const testUser = {
  id: 1,
  username: "test",
  firstName: "test",
  lastName: "test",
  email: "testemail@email.com",
};

const commonBeforeEach = async () => {
  // Mock api calls
  await BorrowcapApi.getLoanRequests.mockResolvedValue([
    {
      id: 2,
      borrowerId: 2,
      amtRequested: "10000",
      purpose: "Education",
      purposeId: 3,
      appOpenDate: "2023/01/01",
      interestRate: "0.07",
      term: 36,
      installmentAmt: "319.86",
    },
  ]);

  await BorrowcapApi.getApprovedRequests.mockResolvedValue([
    {
      id: 1,
      borrowerId: 2,
      amtRequested: "10000",
      amtApproved: "9000",
      amtFunded: "5000",
      purpose: "Car",
      purposeId: 2,
      appOpenDate: "2023/01/01",
      appApprovedDate: "2023/01/05",
      fundingDeadline: "2023/02/05",
      interestRate: "0.05",
      term: 12,
      installmentAmt: "428.04",
      availableForFunding: false,
      isFunded: true,
    },
  ]);
};

module.exports = { testUser, commonBeforeEach };
