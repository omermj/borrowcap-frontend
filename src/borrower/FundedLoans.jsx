import { useState, useEffect, useContext } from "react";
import TableComponent from "../common/TableComponent";
import BorrowcapApi from "../api/api";
import { formatCurrency, formatDate, formatPercent } from "../helpers/format";
import UserContext from "../auth/UserContext";

/** Table showing all Fundend Loans for logged in user */

const FundedLoans = () => {
  const { currentUser } = useContext(UserContext);
  const [fundedLoans, setFundedLoans] = useState([]);

  const payIcon = () => {
    return <i className="bi-credit-card-fill"></i>;
  };

  // get active loan requests on initial render
  useEffect(() => {
    async function fetchFundedLoans() {
      const requests = await BorrowcapApi.getFundedLoansByBorrowerId(
        currentUser.id
      );
      setFundedLoans(requests);
    }
    fetchFundedLoans();
  }, [currentUser.id]);

  const handlePay = async (id) => {
    await BorrowcapApi.payInstallment(id);
    const newReq = await BorrowcapApi.getFundedLoansByBorrowerId(
      currentUser.id
    );
    setFundedLoans([...newReq]);
  };

  const headers = {
    id: { label: "ID", formatter: "none" },
    amtFunded: {
      label: "Amount",
      formatter: formatCurrency,
    },
    fundedDate: { label: "Funded Date", formatter: formatDate },
    interestRate: { label: "Interest Rate", formatter: formatPercent },
    term: { label: "Term", formatter: "none" },
    remainingBalance: { label: "Remaining Balance", formatter: formatCurrency },
    installmentAmt: { label: "Installment", formatter: formatCurrency },
    payBtn: {
      label: "Pay",
      formatter: "button",
      link: "/fundedloans/pay",
      onClick: handlePay,
      icon: payIcon,
    },
  };

  if (!fundedLoans.length) return <div></div>;

  return (
    <div className="border mb-4">
      <div className="py-2">
        <span className="align-middle h5">Funded Loans</span>
      </div>
      <TableComponent headers={headers} tableData={fundedLoans} />
    </div>
  );
};

export default FundedLoans;
