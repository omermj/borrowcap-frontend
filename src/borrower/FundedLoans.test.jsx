import { render, waitFor, screen, fireEvent } from "@testing-library/react";
import FundedLoans from "./FundedLoans.jsx";
import { expect, it } from "vitest";
import UserContext from "../auth/UserContext";
import { commonBeforeEach, testUser } from "./_testCommon.js";
import BorrowcapApi from "../api/api.js";

// Mock api calls
vi.mock("../api/api");
beforeEach(commonBeforeEach);

// Smoke test
it("renders without crashing", () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <FundedLoans />
    </UserContext.Provider>
  );
});

// Snapshot test
it("matches the snapshot", () => {
  const { asFragment } = render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <FundedLoans />
    </UserContext.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

// Test table rendering
it("renders table with funded loans", async () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <FundedLoans />
    </UserContext.Provider>
  );

  // Wait for api call to resolve and populate values
  await waitFor(async () => {
    expect(BorrowcapApi.getFundedLoansByUserId).toHaveBeenCalledWith(
      testUser.id
    );
  });

  // Verify table is populated
  expect(screen.getByText("Amount")).toBeInTheDocument();
  expect(screen.getByText("Funded Date")).toBeInTheDocument();
  expect(screen.getByText("Interest Rate")).toBeInTheDocument();
  expect(screen.getByText("Term")).toBeInTheDocument();
  expect(screen.getByText("Remaining Balance")).toBeInTheDocument();
  expect(screen.getByText("Installment")).toBeInTheDocument();
  expect(screen.getByText("Pay")).toBeInTheDocument();

  // Verify table has correct data
  expect((await screen.findAllByText("$25,000.00")).length).toEqual(2);
  expect(screen.getByText("2023-01-05")).toBeInTheDocument();
  expect(screen.getByText("7.00%")).toBeInTheDocument();
  expect(screen.getByText("36")).toBeInTheDocument();
  expect(screen.getByText("$771.93")).toBeInTheDocument();
});

// Test Fund button
it("handles Fund button click", async () => {
  render(
    <UserContext.Provider value={{ currentUser: testUser }}>
      <FundedLoans />
    </UserContext.Provider>
  );

  // Wait for api call to resolve and populate values
  await waitFor(async () => {
    expect(BorrowcapApi.getFundedLoansByUserId).toHaveBeenCalledWith(
      testUser.id
    );
  });

  // Click Fund button
  async () => {
    const fundButton = screen.getByTestId("pay-button");
    fireEvent.click(fundButton);

    await waitFor(() => {
      expect(BorrowcapApi.payInstallment).toHaveBeenCalled();
      expect(BorrowcapApi.getFundedLoansByUserId).toHaveBeenCalled();
    });
  };
});
