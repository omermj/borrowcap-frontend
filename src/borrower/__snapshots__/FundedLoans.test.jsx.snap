// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches the snapshot 1`] = `
<DocumentFragment>
  <div>
    No funded loans.
  </div>
</DocumentFragment>
`;
